src/instruction_counter.rs

```diff
@@ -0,0 +1,121 @@
+use glob::glob;
+use regex::Regex;
+use std::{path::PathBuf, process::Command};
+
+use haybale::{backend::Backend, Location, State};
+
+pub type Disassem = Vec<String>;
+
+/// Find the bc file to be passed to llc
+fn find_bc_file(bc_dir: &String, board_name: &String) -> String {
+    let pat = format!(r"{}{}*.bc", bc_dir, board_name);
+    let paths = glob(&pat)
+        .unwrap()
+        .map(|x| x.unwrap())
+        .collect::<Vec<PathBuf>>();
+    if paths.len() != 1 {
+        panic!("found multiple dep files");
+    }
+    paths[0].to_str().unwrap().to_owned()
+}
+
+/// Read the output of llc from disc
+fn read_llc_output(mut bc_path: String) -> String {
+    let len = bc_path.len();
+    // change the file extension from ".bc" to ".s"
+    bc_path.truncate(len - 2);
+    bc_path.push_str("s");
+    std::fs::read_to_string(bc_path).expect("could not open llc output")
+}
+
+/// Generate annotated disassembly using llc
+pub fn get_disassembly(bc_dir: &String, board_name: &String) -> Disassem {
+    let bc_path = find_bc_file(bc_dir, board_name);
+
+    let mut llc_command = Command::new("llc-13");
+    llc_command.arg(&bc_path);
+    llc_command.status().expect("llc process failed to execute");
+
+    let llc_output = read_llc_output(bc_path);
+    llc_output.lines().map(|s| s.to_owned()).collect()
+}
+
+/// Build the regexes used to find the function and basic block location
+fn build_func_and_bb_patterns(location: &Location) -> (Regex, Regex) {
+    // matches the start of the desired function
+    let func_name = &location.func.name;
+    let func_pat = format!(r"^{}:$", func_name);
+    let func_re = Regex::new(&func_pat).unwrap();
+
+    // matches the start of the desired bb
+    // TODO: The basic block name can take on many forms...
+    let bb_name = &location.bb.name;
+    let bb_pat = format!(r"^.*(@\s*{}).*$", bb_name);
+    let bb_re = Regex::new(&bb_pat).unwrap();
+
+    (func_re, bb_re)
+}
+
+/// Count the number of machine instructions corresponding to the current path
+pub fn count_instructions<'p, B: Backend>(
+    disassembly: &Disassem,
+    state: &State<'p, B>,
+) -> Result<(String, usize), String> {
+    // matches any line that is a machine instruction
+    let instr_re = Regex::new(r"^(\s*)([^@_\s\.])(.*)$").unwrap();
+    // matches the start of a function or bb
+    let bb_or_func_re = Regex::new(r"(^_.+:$)|(^@\s*%bb\.\d+:.*$)").unwrap();
```

Comment:
If we are counting instructions along a lot of paths we may want to compile these regex's once using lazy_static (check out the regex docs), but this is fine for now. Compiling is much slower than matching so it saves a reasonable amount of time to only compile each regex once.